src/services/db-service/repositories/userRepository.ts

```diff
@@ -1,13 +1,52 @@
 import { Injectable } from '@nestjs/common';
 
+import { Collection } from 'mongodb';
+
 import { MongoService } from '../mongoService';
 import { User } from '../../../models/User';
+import { WithoutId } from '../../../typeUtils';
+import { modelSchemas } from '../../../models';
 
 @Injectable()
 export class UserRepository {
-  constructor(private dbService: MongoService) {}
+  constructor(private readonly dbService: MongoService<'User'>) {}
 
   createUser(user: Omit<User, 'id'>): Promise<User> {
     return this.dbService.createOne('User', user);
   }
+
+  createOrUpdateUser(
+    filter: Partial<WithoutId<User>>,
+    update: Partial<WithoutId<User>>,
+  ): Promise<User> {
+    return this.dbService.upsert('User', filter, update);
+  }
+
+  handleLogin(createUserArgs: {
+    lichessPuzzleRating: number;
+    lichessUsername: string;
+    lichessId: string;
+  }) {
+    this._collection()
+      .findOneAndUpdate(
+        { lichessId: createUserArgs.lichessId },
+        {
+          $set: { ...createUserArgs },
+          $setOnInsert: {
+            username: createUserArgs.lichessUsername,
```

Comment:
If they are logging in for the first time then create a username thats equal tot he lichess username